libs/types/src/pools.rs

```diff
@@ -0,0 +1,227 @@
+pub struct PoolParameters {
+	/// Minimum duration for an epoch.
+	pub min_epoch_time: Moment,
+	/// Maximum time between the NAV update and the epoch closing.
+	pub max_nav_age: Moment,
+}
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct PoolChanges<Rate, MaxTokenNameLength, MaxTokenSymbolLength, MaxTranches>
+where
+	MaxTokenNameLength: Get<u32>,
+	MaxTokenSymbolLength: Get<u32>,
+	MaxTranches: Get<u32>,
+{
+	pub tranches: Change<BoundedVec<TrancheUpdate<Rate>, MaxTranches>>,
+	pub tranche_metadata:
+		Change<BoundedVec<TrancheMetadata<MaxTokenNameLength, MaxTokenSymbolLength>, MaxTranches>>,
+	pub min_epoch_time: Change<Moment>,
+	pub max_nav_age: Change<Moment>,
+}
+
+/// Information about the deposit that has been taken to create a pool
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, Default, TypeInfo)]
+pub struct PoolDepositInfo<AccountId, Balance> {
+	pub depositor: AccountId,
+	pub deposit: Balance,
+}
+
+/// The core metadata about the pool which we can attach to an event
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct PoolEssence<
+	CurrencyId,
+	Balance,
+	TrancheCurrency,
+	Rate,
+	MaxTokenNameLength,
+	MaxTokenSymbolLength,
+> where
+	CurrencyId: Copy,
+	MaxTokenNameLength: Get<u32>,
+	MaxTokenSymbolLength: Get<u32>,
+{
+	/// Currency that the pool is denominated in (immutable).
+	pub currency: CurrencyId,
+	/// The maximum allowed reserve on a given pool
+	pub max_reserve: Balance,
+	/// Maximum time between the NAV update and the epoch closing.
+	pub max_nav_age: Moment,
+	/// Minimum duration for an epoch.
+	pub min_epoch_time: Moment,
+	/// Tranches on a pool
+	pub tranches:
+		Vec<TrancheEssence<TrancheCurrency, Rate, MaxTokenNameLength, MaxTokenSymbolLength>>,
+}
+
+impl<CurrencyId, TrancheCurrency, EpochId, Balance, Rate, MetaSize, Weight, TrancheId, PoolId>
+	PoolDetails<
+		CurrencyId,
+		TrancheCurrency,
+		EpochId,
+		Balance,
+		Rate,
+		MetaSize,
+		Weight,
+		TrancheId,
+		PoolId,
+	> where
+	Balance: FixedPointOperand + BaseArithmetic + Unsigned + From<u64>,
+	CurrencyId: Copy,
+	EpochId: BaseArithmetic,
+	MetaSize: Get<u32> + Copy,
+	PoolId: Copy + Encode,
+	Rate: FixedPointNumber<Inner = Balance>,
+	TrancheCurrency: Copy + cfg_traits::TrancheCurrency<PoolId, TrancheId>,
+	TrancheId: Clone + From<[u8; 16]> + PartialEq,
+	Weight: Copy + From<u128>,
+{
+	pub fn start_next_epoch(&mut self, now: Moment) -> DispatchResult {
+		self.epoch.current += One::one();
+		self.epoch.last_closed = now;
+		// TODO: Remove and set state rather to EpochClosing or similar
+		// Set available reserve to 0 to disable originations while the epoch is closed but not executed
+		self.reserve.available = Zero::zero();
+
+		Ok(())
+	}
+
+	pub fn execute_previous_epoch(&mut self) -> DispatchResult {
+		self.reserve.available = self.reserve.total;
+		self.epoch.last_executed += One::one();
+		Ok(())
+	}
```

Comment:
Can you create a ticket for us to keep track that we need to check, how we can ensure, that pools maintains all invariants?

Reply:
https://github.com/centrifuge/centrifuge-chain/issues/1091